src/Components/Web.JS/src/Boot.Server.ts

```diff
@@ -122,6 +122,63 @@ async function initializeConnection(options: CircuitStartOptions, logger: Logger
 
   Blazor._internal.forceCloseConnection = () => connection.stop();
 
+  Blazor._internal.sendJSDataStream = async (data: ArrayBufferView, streamId: string, maximumIncomingBytes: number) => {
+    // Run the rest in the background, without delaying the completion of the call to sendJSDataStream
+    // otherwise we'll deadlock (.NET can't begin reading until this completes, but it won't complete
+    // because nobody's reading the pipe)
+    setTimeout(async () => {
+      const maxMillisecondsBetweenAcks = 500;
+      let numChunksUntilNextAck = 5;
+      let lastAckTime = new Date().valueOf();
+      try {
+        // Chunk Size is the max SignalR message size less 512 bytes assumed overhead.
+        const chunkSize = maximumIncomingBytes - 512;
+        if (chunkSize <= 0) {
+          throw new Error('The chunk size must be greater than 0.')
+        }
+
+        let position = 0;
+
+        // Note: The server-side `StreamBufferCapacity` option (defaults to 10) can be configured to limit how many
+        // stream items from the client (per stream) will be stored before reading any more stream items (thus applying backpressure).
+        while (position < data.byteLength) {
+          const nextChunkSize = Math.min(chunkSize, data.byteLength - position);
+          console.log(`Sending a chunk of length ${nextChunkSize}`);
+          const nextChunkData = new Uint8Array(data.buffer, data.byteOffset + position, nextChunkSize);
+
+          numChunksUntilNextAck--;
+          if (numChunksUntilNextAck > 1) {
+            // Most of the time just send and buffer within the network layer
+            await connection.send('ReceiveJSDataChunk', streamId, nextChunkData, null);
+          } else {
+            // But regularly, wait for an ACK, so other events can be interleaved
+            // The use of "invoke" (not "send") here is what prevents the JS side from queuing up chunks
+            // faster than the .NET side can receive them. It means that if there are other user interactions
+            // while the transfer is in progress, they would get inserted in the middle, so it would be
+            // possible to navigate away or cancel without first waiting for all the remaining chunks.
+            const streamIsAlive = await connection.invoke<boolean>('ReceiveJSDataChunk', streamId, nextChunkData, null);
```

Comment:
Could this be done in the same way we do with render batches instead? where the server explicitly sends an ack?

What would be the problem with the client sending 1, 2, 3, 4, 5 and assuming the server receives them and they are ready when a read operation happens and consumes them, sends back a single 5 ack indicating the number of consumed segments, which enables the client to send 5 more, instead of waiting for individual acks after the first 5 chunks.

Reply:
Not sure I follow. 

We essentially are doing the method you're suggesting (sending chunks and waiting for an ack after every `numChunksUntilNextAck` chunks).

> instead of waiting for individual acks after the first 5 chunks.

We aren't waiting for individual acks after each chunk (not sure if that's what you meant here).

Reply:
Yeah, I see now better how this works.

Reply:
@javiercn Are you OK with the descriptiveness of the comments? Elsewhere there are notes that we should clarify more how it works, but reading the comment here now, it seems reasonably clear. Not sure if/how we should make it clearer.

Reply:
ping @javiercn safe to close?